contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,1422 @@
+    }
+
+    /// @notice Copies delegation from the legacy KEEP staking contract to T
+    ///         staking contract. No tokens are transferred. Caches the active
+    ///         stake amount from KEEP staking contract. Can be called by
+    ///         anyone.
+    function stakeKeep(address operator) external override {
+        require(operator != address(0), "Operator must be specified");
+        OperatorInfo storage operatorStruct = operators[operator];
+
+        require(
+            operatorStruct.owner == address(0),
+            "Operator is already in use"
+        );
+
+        uint96 tAmount = getKeepAmountInT(operator);
+        require(tAmount != 0, "Nothing to sync");
+
+        operatorStruct.keepInTStake = tAmount;
+        operatorStruct.owner = keepStake.resolveOwner(
+            keepStakingContract,
+            operator
+        );
+        operatorStruct.authorizer = keepStakingContract.authorizerOf(operator);
+        operatorStruct.beneficiary = keepStakingContract.beneficiaryOf(
+            operator
+        );
+        emit OperatorStaked(
+            StakeType.KEEP,
+            operatorStruct.owner,
+            operator,
+            operatorStruct.beneficiary,
+            operatorStruct.authorizer,
+            tAmount
+        );
+    }
+
+    /// @notice Copies delegation from the legacy NU staking contract to T
+    ///         staking contract, additionally appointing beneficiary and
+    ///         authorizer roles. Caches the amount staked in NU staking
+    ///         contract. Can be called only by the original delegation owner.
+    function stakeNu(
+        address operator,
+        address payable beneficiary,
+        address authorizer
+    ) external override {
+        require(
+            operator != address(0) &&
+                beneficiary != address(0) &&
+                authorizer != address(0),
+            "Roles must be specified"
+        );
+        OperatorInfo storage operatorStruct = operators[operator];
+        (, uint256 createdAt, ) = keepStakingContract.getDelegationInfo(
+            operator
+        );
+        require(
+            createdAt == 0 && operatorStruct.owner == address(0),
+            "Operator is already in use"
+        );
+
+        uint96 tAmount = getNuAmountInT(msg.sender, operator);
+        require(tAmount > 0, "Nothing to sync");
+
+        operatorStruct.nuInTStake = tAmount;
+        operatorStruct.owner = msg.sender;
+        operatorStruct.authorizer = authorizer;
+        operatorStruct.beneficiary = beneficiary;
+
+        emit OperatorStaked(
+            StakeType.NU,
+            msg.sender,
+            operator,
+            operatorStruct.beneficiary,
+            operatorStruct.authorizer,
+            tAmount
+        );
+    }
+
+    /// @notice Extract owner of Keep stake from grantee in ManagedGrant
+    function refreshKeepManagedGrantOwner(address operator)
+        external
+        onlyOwnerOrOperator(operator)
```

Comment:
Yeah, good catch. But I think `onlyOwnerOrOperator` still does not solve the problem because operators are now able to take over stakes. 

I think the root of the problem lies in `keepStake.resolveOwner` call: `refreshKeepManagedGrantOwner` aims at refreshing just managed grant owners but `keepStake.resolveOwner` is also resolving liquid token owners. The following scenario is possible:
- in T staking contract 0xABC delegates their liquid T to 0xDEF,
- malicious operator 0xDEF delegates in Keep staking contract liquid KEEP from 0xDEF to 0xDEF,
- malicious operator 0xDEF calls `refreshKeepManagedGrantOwner` and they are becoming the owner of 0xABC's stake.

Also, I've just realized Keep staking contract allows liquid token owners [to transfer their stake ownership](https://github.com/keep-network/keep-core/blob/68bfbe2c2deef66ac2e01e904b13ab3a7d576614/solidity-v1/contracts/TokenStaking.sol#L469-L473). This is something we had to include in staking contract when were deploying new staking contract version and borrowed company tokens to existing stakers so that they do not have to suffer undelegation period in the old staking contract version. Not widely used functionality but it is still there and one can call it...

So I think `refreshKeepManagedGrantOwner` should really become `refreshKeepStakeOwner`, especially that `keepStake.resolveOwner` called in this function resolves all types of owners. And, we need to ensure stealing stakes is not possible, so what if we `require` that `operatorInfo.startTStakingTimestamp > createdAt`, while `createdAt` comes from `keepStakingContract.getDelegationInfo`?

Then, we can remove the modifier from this function and let it be called by anyone so that if someone has lost their old owner access and earlier forgot to call this function, they do not need to count on operator for doing so. I've seen a lot of weird situations already...

(still need to think about it but posting early so that we can both can 🧠 )

Reply:
but `operatorInfo.startTStakingTimestamp` is only for T stake. Or I'm missing something?
Also for me `refreshKeepStakeOwner` looks similar to top-ups, changing stake info. So it looks natural to have `onlyOwnerOrOperator`

Reply:
> but `operatorInfo.startTStakingTimestamp` is only for T stake. Or I'm missing something?

What if we make it `startStakingTimestamp`? This change will not allow legacy KEEP/NU stakers with no T to undelegate earlier than 24h but I think it's fine. As a staker, I'd be much more comfortable about it than about the fact operator can steal my tokens.

Reply:
I'm not convienced. For Keep/Nu there is no much sense to store it because main problem with unstaking is to delegate to another operator which is not really possible to do. And this will add additional gas cost to store one more slot for stakeKeep and stakeNu when most operators will not be affected by issues with `refreshKeepStakeOwner `

Reply:
> For Keep/Nu there is no much sense to store it because main problem with unstaking is to delegate to another operator which is not really possible to do.

I am afraid I do not follow this part. Let me try to rephrase the problem I see:
- On Keep side it is possible the owner is changed.
- Given that we cache Keep owner in T staking contract, we need to provide an option to refresh the cached value.
- Right now, any operator can delegate 1 liquid KEEP in Keep staking contract, then call refresh function and become owner of the stake. They are now both owner and operator and the real owner can do nothing about it.
- This affects liquid T stakes and legacy Nu stakes. This does not affect legacy Keep stakes because operator address can not be reused in Keep staking contract.

I think the only validation we can do is to ensure the delegation in Keep staking contract was not created after the delegation in T staking contract, no matter if it was liquid T delegation or legacy Nu delegation.

Reply:
> Also for me `refreshKeepStakeOwner` looks similar to top-ups, changing stake info. So it looks natural to have `onlyOwnerOrOperator`

I do not disagree in principle. I am just worried about real-world situations we already experienced when some grantees just lose access to the original grantee account after some time. I've also seen some operators dropping entire droplets, including their keys.

If we can make this function safe and ensure always the right owner is resolved, I think we should remove the ACL. If we can not... then we are probably having another problem, like the one described in the previous comment.

Reply:
> * Right now, any operator can delegate 1 liquid KEEP in Keep staking contract, then call refresh function and become owner of the stake. They are now both owner and operator and the real owner can do nothing about it.

Only if we don't have ACL, right?

> I am just worried about real-world situations we already experienced when some grantees just lose access to the original grantee account after some time. I've also seen some operators dropping entire droplets, including their keys.

But for other stakers in that situation - they can't do anything. I see that adding more code just because of some set of irresponsible grantees it's not honest for all other stakers who will pay for this during staking

Reply:
> Only if we don't have ACL, right?

Even if we have the ACL, the operator to which you delegated your stake can steal it. That's why I say everyone is affected, and the additional cost is justified because it's for their own protection, not just for irresponsible grantees.

Reply:
We synced about it with @vzotova and decided to protect against this situation by authorizing only the old stake owner to call this function. The reasoning is:
- old stake owner is required to call `transferStakeOwnership` on Keep staking contract,
- old grantee is required to request grantee change on `ManagedGrant`,
- we need to treat `KeepStake` as trusted, otherwise `stakeKeep` can not be trusted as well.

Reply:
✔️